### PR TITLE
Improve net.Read/WriteTable

### PR DESCRIPTION
**Pull request by meepdarknessmeep**
I accidentally deleted other repository and couldn't update it and since i found a few bugs, i remade it.

### CODE REVIEW

**Comment by thegrb93** [comment]
You need 16 bits for Garrysmod networked entities. I don't think 2048 limit is the same for Garrysmod. Even if it was, you should use 11 bits instead of 12.

**Comment by thegrb93** [comment]
Its sort of nice that you made it not send an entity if it is invalid, but that means every entity that is valid sends an extra bit. I think the number of valid entities networked far outweighs invalid entities so I don't think that is an improvement.

File: garrysmod/lua/includes/extensions/net.lua [hunk 1]
```diff
@@ -179,3 +144,251 @@ function net.ReadType( typeid )
 
 	error( "net.ReadType: Couldn't read type " .. typeid )
 end
+
+local reading, writing
+
+
+local _type = type
+local function type(x)
```
Comment (thegrb93) [inline garrysmod/lua/includes/extensions/net.lua]: Why

**Comment by meepdarknessmeep** [comment]
you say it's 16 bits, but CClientEntityList::GetMaxEntities() returns 8096 (1/2 is always non networkable, thus leading to 12 bits
http://nl.cm/xsf

**Comment by thegrb93** [comment]
I'll support that if the devs can verify that. Still I don't think sending that extra boolean is worth it.

**Comment by meepdarknessmeep** [comment]
So you'd rather have invalid/non-networkable entities sent as the world entity instead of NULL entities?
That seems like a bad idea imo. Imagine what people would encounter and have no idea why it's happening. It's not about having the least amount of networking possible, it's more about have the best combination of network load/accuracy

**Comment by thegrb93** [comment]
Doesn't EntIndex throw an error if it tries to use an invalid entity? Just do that.

**Comment by meepdarknessmeep** [comment]
That's a terrible idea.
NULL Entities exist and we should acknowledge that. there's even a global variable called NULL that is a null entity.

**Comment by thegrb93** [comment]
Okay that's fine and all, but you shouldn't send them to net.WriteEntity

**Comment by meepdarknessmeep** [comment]
Closing this for now. I might come back at one point to improve it on default gmod again.

**Pull request closed without merging**